jctools-core/src/main/java/org/jctools/queues/MpscRelaxedArrayQueue.java

```diff
@@ -0,0 +1,686 @@
+            this.cycleLengthLog2);
+        if (producerPosition == consumerPosition)
+        {
+            return null;
+        }
+        else
+        {
+            E e;
+            while ((e = lvElement(buffer, offset)) == null)
+            {
+
+            }
+            //can be used a memory_order_relaxed set here, because the consumer position write release the buffer value
+            spElement(buffer, offset, null);
+            //consumer position allows the producers to move the claim limit (aka reduce backpressure)
+            //hence can be set only after the buffer slot release
+            soConsumerPosition(consumerPosition + 1);
+            return e;
+        }
+    }
+
+    @Override
+    public E peek()
+    {
+        final long consumerPosition = lpConsumerPosition();
+        final long offset = CircularArrayOffsetCalculator.calcElementOffset(consumerPosition, this.mask);
+        E e = lvElement(this.buffer, offset);
+        if (e != null)
+        {
+            return e;
+        }
+        final int activeCycleIndex = lvActiveCycleIndex();
+        final long producerCycleClaim = UnsafeLongArrayAccess.lvValue(this.producerCycleClaim, activeCycleIndex);
+        final long producerPosition = producerPosition(
+            cycleId(producerCycleClaim, this.cycleIdBitShift),
+            positionOnCycle(producerCycleClaim, this.positionOnCycleMask),
+            this.cycleLengthLog2);
+        if (producerPosition == consumerPosition)
+        {
+            return null;
+        }
+        else
+        {
+            while ((e = lvElement(buffer, offset)) == null)
+            {
+
+            }
+            return e;
+        }
+    }
+
+    @Override
+    public int size()
+    {
+        long after = lvConsumerPosition();
+        int positionOnCycle;
+        long producerClaim;
+        long before;
+        do
+        {
+            before = after;
+            final int activeClaim = lvActiveCycleIndex();
+            producerClaim = UnsafeLongArrayAccess.lvValue(this.producerCycleClaim, activeClaim);
+            after = lvConsumerPosition();
+            positionOnCycle = positionOnCycle(producerClaim, this.positionOnCycleMask);
+        }
+        while (positionOnCycle > this.cycleLength || before != after);
+        //need to have a stable consumer and a valid claim (into the cycle)
+        final long size =
+            producerPosition(cycleId(producerClaim, this.cycleIdBitShift), positionOnCycle, this.cycleLengthLog2) -
+                after;
+        if (size > Integer.MAX_VALUE)
+        {
+            return Integer.MAX_VALUE;
+        }
+        else
+        {
+            return (int) size;
+        }
+    }
+
+    @Override
+    public void clear()
+    {
+        while (poll() != null)
+        {
+            // if you stare into the void
+        }
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+        return size() == 0;
```

Comment:
Should be:
```
       final long producerPosition = producerPosition(
            cycleId(producerCycleClaim, this.cycleIdBitShift),
            positionOnCycle(producerCycleClaim, this.positionOnCycleMask),
            this.cycleLengthLog2);
        return (producerPosition == consumerPosition);
```
?

Reply:
good point!!!

Reply:
Implementing it like this in the atomic version of the q:
```
    @Override
    public boolean isEmpty()
    {
        final long consumerPosition = lvConsumerPosition();
        final int activeCycleIndex = lvActiveCycleIndex();
        final long producerCycleClaim = AtomicLongArrayAccess.lvValue(this.producerCycleClaim, activeCycleIndex);
        final long producerPosition = producerPosition(
            cycleId(producerCycleClaim, this.cycleIdBitShift),
            positionOnCycle(producerCycleClaim, this.positionOnCycleMask),
            this.cycleLengthLog2);
        return (producerPosition == consumerPosition);
    }
```
It will cause a fail int the sanity test:
```
java.lang.AssertionError: Observed no element in non-empty queue 
Expected :0
Actual   :350
 <Click to see difference>


	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.jctools.queues.QueueSanityTest.testPollAfterIsEmpty(QueueSanityTest.java:361)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:51)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

Reply:
@nitsanw 
Turning it into this seems to fix that:
```
    @Override
    public boolean isEmpty()
    {
        final AtomicLongArray producerCycleClaims = this.producerCycleClaim;
        final int cycleLength = this.cycleLength;
        final int positionOnCycleMask = this.positionOnCycleMask;
        final int cycleIdBitShift = this.cycleIdBitShift;
        final int cycleLengthLog2 = this.cycleLengthLog2;
        final long consumerPosition = lvConsumerPosition();
        int activeCycleIndex;
        long producerCycleClaim;
        int positionOnCycle;
        do
        {
            activeCycleIndex = lvActiveCycleIndex();
            producerCycleClaim = AtomicLongArrayAccess.lvValue(producerCycleClaims, activeCycleIndex);
        }
        while ((positionOnCycle = positionOnCycle(producerCycleClaim, positionOnCycleMask)) > cycleLength);

        final long producerPosition = producerPosition(
            cycleId(producerCycleClaim, cycleIdBitShift),
            positionOnCycle,
            cycleLengthLog2);
        return (producerPosition == consumerPosition);
    }
```
Makes sense?
But TBH I'm not seeing such a big improvement vs the original implementation...